src/main/java/com/fasterxml/jackson/databind/deser/std/EnumDeserializer.java

```diff
@@ -253,6 +272,8 @@ protected Object _fromString(JsonParser p, DeserializationContext ctxt,
     {
         CompactStringObjectMap lookup = ctxt.isEnabled(DeserializationFeature.READ_ENUMS_USING_TO_STRING)
                 ? _getToStringLookup(ctxt) : _lookupByName;
+        lookup = _hasEnumNaming(ctxt) ? _getEnumNamingLookup(ctxt) : lookup;
```

Comment:
Ok, getting better, but I think all of introspection really should be done when constructing `EnumDeserializer`, and not dynamically. That avoids having to use synchronization and we just replace names used when constructing/initializing deserializer.